File: src/gate.mjs
// §8 Gate de calidad. Codigo puro: tres notas y sus incidencias entran, una
// decision sale. El gate no sabe si las revisiones vinieron de una llamada al
// validador o de tres (§5), solo que ya pasaron VD-10.
import { DIMENSIONES } from './esquemas.mjs';

export function notas(revisiones) {
  return revisiones.map((r) => r.nota);
}

export function media(valores) {
  if (valores.length === 0) return 0;
  return valores.reduce((a, b) => a + b, 0) / valores.length;
}

export function incidenciasGraves(revisiones) {
  return revisiones.flatMap((r) => (r.incidencias ?? [])
    .filter((i) => i.severidad === 'grave')
    .map((i) => ({ ...i, dimension: r.dimension })));
}

/**
 * aprueba = min(notas) >= nota_minima y media(notas) >= media_minima
 *           y ninguna incidencia grave.
 * La incidencia grave veta por si sola: un capitulo puede sacar tres cuatros y
 * caer por una contradiccion de canon, porque eso no se arregla puntuando mas alto.
 */
export function gate(revisiones, configGate) {
  const ns = notas(revisiones);
  const minima = Math.min(...ns);
  const promedio = media(ns);
  const graves = incidenciasGraves(revisiones);

  const motivos = [];
  if (minima < configGate.nota_minima) {
    motivos.push(`nota minima ${minima} por debajo de ${configGate.nota_minima}`);
  }
  if (promedio < configGate.media_minima) {
    motivos.push(`media ${promedio.toFixed(2)} por debajo de ${configGate.media_minima}`);
  }
  if (graves.length) {
    motivos.push(`${graves.length} incidencia(s) grave(s)`);
  }

  return {
    aprueba: motivos.length === 0,
    minima,
    media: promedio,
    graves,
    motivos,
  };
}

/** Al agotar intentos se conserva el de mejor media, en estado propuesto (§8). */
export function mejorIntento(intentos) {
  const conRevisiones = intentos.filter((i) => Array.isArray(i.revisiones) && i.revisiones.length);
  if (!conRevisiones.length) return intentos.at(-1) ?? null;
  return conRevisiones.reduce((mejor, actual) => (
    media(notas(actual.revisiones)) > media(notas(mejor.revisiones)) ? actual : mejor
  ));
}

/** Las incidencias del intento anterior, ordenadas por severidad (§8). */
export function incidenciasOrdenadas(revisiones) {
  const orden = { grave: 0, aviso: 1 };
  return revisiones
    .flatMap((r) => (r.incidencias ?? []).map((i) => ({ ...i, dimension: r.dimension })))
    .sort((a, b) => (orden[a.severidad] ?? 9) - (orden[b.severidad] ?? 9)
      || DIMENSIONES.indexOf(a.dimension) - DIMENSIONES.indexOf(b.dimension));
}

File: novela/gate.py
# §8 Gate de calidad. Codigo puro: tres notas y sus incidencias entran, una
# decision sale. El gate no sabe si las revisiones vinieron de una llamada al
# validador o de tres (§5), solo que ya pasaron VD-10.
from .esquemas import DIMENSIONES
from .util import redondear


def notas(revisiones):
    return [r['nota'] for r in revisiones]


def media(valores):
    if not valores:
        return 0
    return sum(valores) / len(valores)


def incidencias_graves(revisiones):
    graves = []
    for r in revisiones:
        for i in r.get('incidencias') or []:
            if i.get('severidad') == 'grave':
                graves.append({**i, 'dimension': r['dimension']})
    return graves


def gate(revisiones, config_gate):
    """aprueba = min(notas) >= nota_minima y media(notas) >= media_minima
    y ninguna incidencia grave.

    La incidencia grave veta por si sola: un capitulo puede sacar tres cuatros y
    caer por una contradiccion de canon, porque eso no se arregla puntuando mas
    alto.
    """
    ns = notas(revisiones)
    minima = min(ns) if ns else 0
    promedio = media(ns)
    graves = incidencias_graves(revisiones)

    motivos = []
    if minima < config_gate['nota_minima']:
        motivos.append('nota minima {} por debajo de {}'.format(
            minima, config_gate['nota_minima']))
    if promedio < config_gate['media_minima']:
        motivos.append('media {:.2f} por debajo de {}'.format(
            redondear(promedio, 2), config_gate['media_minima']))
    if graves:
        motivos.append('{} incidencia(s) grave(s)'.format(len(graves)))

    return {
        'aprueba': not motivos,
        'minima': minima,
        'media': promedio,
        'graves': graves,
        'motivos': motivos,
    }


def mejor_intento(intentos):
    """Al agotar intentos se conserva el de mejor media, en estado propuesto (§8)."""
    con_revisiones = [
        i for i in intentos
        if isinstance(i.get('revisiones'), list) and i['revisiones']
    ]
    if not con_revisiones:
        return intentos[-1] if intentos else None

    mejor = con_revisiones[0]
    for actual in con_revisiones[1:]:
        if media(notas(actual['revisiones'])) > media(notas(mejor['revisiones'])):
            mejor = actual
    return mejor


def incidencias_ordenadas(revisiones):
    """Las incidencias del intento anterior, ordenadas por severidad (§8)."""
    orden = {'grave': 0, 'aviso': 1}
    todas = []
    for r in revisiones:
        for i in r.get('incidencias') or []:
            todas.append({**i, 'dimension': r['dimension']})

    def clave(incidencia):
        dimension = incidencia.get('dimension')
        return (
            orden.get(incidencia.get('severidad'), 9),
            DIMENSIONES.index(dimension) if dimension in DIMENSIONES else len(DIMENSIONES),
        )

    return sorted(todas, key=clave)
